### PR TITLE
DRAFT: Improve documentation across the entire project

### PR DESCRIPTION
**Pull request by hatf0**

### CODE REVIEW

**Comment by hatf0** [comment]
I'm thinking of extracting out the massive collection of examples out to the `examples/` folder, just because I believe they're currently a little hard to follow. 

Also, @9il, what does the `annotated` flag do (w.r.t. `deserializeIon`)?

**Comment by 9il** [comment]
> Also, @9il, what does the `annotated` flag do (w.r.t. `deserializeIon`)?

It is adds addtional `IonAnnotatoins` argument. The problem is that we may wan't to handle the first one annotation while the following one annotations are related to the value itself. So, this flags helps to solve annotation encapsulation.

File: examples/basic-attributes/source/app.d [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	string field1;
+
+	@serdeOptional
+	int field2;
+}
+
+// A simple struct with all required members
+struct RequiredMembers
+{
+	@serdeRequired
```
Comment (9il) [inline examples/basic-attributes/source/app.d:24]: Members are required by default. This attribute is redundant.

File: examples/basic-attributes/source/app.d [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	// For the sake of clarity, all examples below will be in the Ion text format. However,
+	// they can just as easily be translated to the Ion binary format.
+
+	/++++++++++++ Optional UDA ++++++++++++/
+	// An optional UDA means that the member does not have to always be present,
+	// and the deserializer will not throw if it is not present.
+
+	// Deserialize an instance of OptionalMembers with all members present
+	{
+		static immutable data = q"{
```
Comment (9il) [inline examples/basic-attributes/source/app.d:42]: It is better to use `` or q{}. q"" doesn't look nice.

File: examples/basic-attributes/source/app.d [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+			{
+				"field1": "Foobar",
+				"field2": 123
+			}
```
Comment (9il) [inline examples/basic-attributes/source/app.d:46]: One line definition looks better better.

File: examples/basic-attributes/source/app.d [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	@serdeRequired
+	string field1;
+
+	@serdeRequired
+	int field2;
+}
+
+void main()
+{
+	// For the sake of clarity, all examples below will be in the Ion text format. However,
```
Comment (9il) [inline examples/basic-attributes/source/app.d:33]: Please use 4 spaces, not tabs

File: examples/basic-attributes/source/app.d [hunk 5]
```diff
@@ -0,0 +1,102 @@
+import std.stdio;
+// Provides the text deserialization functions
+import mir.ion.deser.text;
+// Provides the binary deserialization functions
+import mir.ion.deser.ion;
+// Provides the functions to go between formats (binary -> text, text -> binary, etc)
+import mir.ion.conv;
+// Provides all of the UDAs that Mir Ion recognizes (and uses)
+import mir.serde;
```
Comment (9il) [inline examples/basic-attributes/source/app.d:9]: Selective imports are desired for examples, it makes clear where the definition can be found.

**Comment by 9il** [comment]
- The pivot table of UDAs, even with empty descriptions
- binary literals
- CLOB and BLOB support (do we already support CLOBs?)

have priority over examples

**Review by 9il — CHANGES_REQUESTED** [review]

**Comment by hatf0** [comment]
> * The pivot table of UDAs, even with empty descriptions
> * binary literals
> * CLOB and BLOB support (do we already support CLOBs?)
> 
> have priority over examples

Understood, I’ll focus on those then. We also already support clobs, it’s just blob support that’s missing (due to no base64 decoding support in Mir Algorithm).

**Comment by 9il** [comment]
> due to no base64 decoding support in Mir Algorithm

Please add it
